### PR TITLE
Revoke Installation

### LINKED ISSUE #21: Contact Operations – Registry – RevokeInstallation
### Discussed in https://github.com/xmtp/xps-gateway/discussions/14

<div type='discussions-op-text'>

<sup>Originally posted by **jac18281828** December 22, 2023</sup>
Initial discussion is [here](https://github.com/xmtp/libxmtp/discussions/274)

Device installation is supported by [GrantInstallation](https://github.com/xmtp/xps-gateway/discussions/13).  This may be reversed by the endpoint described below.

#### Request:

- **Method:** `POST`
- **URL:** `/rpc/v1/revokeInstallation`
- **Headers:**
  - `Content-Type: application/json`
- **Body:**
  - **JSON Object:** 
    - `jsonrpc`: `"2.0"`
    - `method`: `"revokeInstallation"`
    - `params`: Array (optional parameters as required)
    - `id`: Request identifier (integer or string)

#### Description
The `revokeInstallation` endpoint is responsible for removing the contact bundle for the XMTP device installation.   The request must be made to a valid did with an XMTP profile.

#### Request
The request for this endpoint should contain the necessary information to authenticate and validate the installation request including the wallet signed payload

##### Parameters:
- `DID` (string): Unique XMTP identifier for the user requesting the revocation.
- `name` (bytes32): Unique identifier naming bundled contents.  **NOTE** 32 byte maximum
- `value` (bytes): Installation bundle bytes payload
- `V` (int): signature V
- `R` (bytes): signature R
- `S` (bytes): signature S

##### Example Request:
```json
{
  "jsonrpc": "2.0",
  "method": "revokeInstallation",
  "params": {
    "did": "12345",
    "name": "xmtp/contact_installation",
    "value": "#######",
    "signature": {
      "V": "12345",
      "R": "valueR",
      "S": "valueS"
    }
  },
  "id": 1
}
```

#### Response
The response will indicate whether the installation is revoked and may include additional information or instructions.

##### Result Fields:
- `status` (string): The status of the request, e.g., 'completed'.
- `message` (string, optional): Additional information or reason for the decision.
- `tx` (string, optional): transaction receipt

##### Example Response:
```json
{
  "jsonrpc": "2.0",
  "result": {
    "status": "completed",
    "message": "Installation revoked.",
    "tx": "<receipt>"
  },
  "id": 1
}
```

#### Error Handling
In case of an error, the response will include an error object with details.

##### Error Object Fields:
- `code` (integer): Numeric code representing the error type.
- `message` (string): Description of the error.

##### Example Error Response:
```json
{
  "jsonrpc": "2.0",
  "error": {
    "code": 403,
    "message": "User not authorized for installation."
  },
  "id": 1
}
```

#### Security Considerations
All requests to `revokeInstallation` must be made over a secure channel. Ensure that user and application IDs are validated, and proper authentication mechanisms are in place to prevent unauthorized requests.
</div>

### PR DESCRIPTION
**Pull request by insipx**
Closes #21 

Has a companion https://github.com/xmtp/didethresolver/pull/35 , to add the logic of signing attributes/delegates/owners

changes the `name` parameter of the RPC to a `XmtpAttribute`, which is able to convert between the string representation of the attribute (`"xmtp/installation/hex"`) and the byte `[u8; 32]` representation, which is taken directly from `lib_didethresolver` so that if the library type changes, it will be evident in the gateway via compilation error.

Other misc changes:
- adds anvil for integration testing (required to revoke attributes w/ eth node)
- only runs lib and doctests for gh pages
- modifies dockerfile as necessary
- runs coverage with trace logs, so we don't get missing coverage errors for those logs anymore

### CODE REVIEW

**Review by jac18281828 — COMMENTED** [review]
Looks like a good start!

I think the key type makes sense.   I think there may be some missed opportunities to add test coverage here too.

**Comment by insipx** [comment]
> Looks like a good start!
> 
> I think the key type makes sense. I think there may be some missed opportunities to add test coverage here too.

yeah definitely still missing tests!

File: xps-gateway/src/rpc/api.rs [hunk 1]
```diff
@@ -12,6 +14,116 @@ pub trait Xps {
...
+    /// removes the contact bundle for the XMTP device installation. Request must be made to a
+    /// valid DID with an XMTP profile.
+    ///
+    /// # Arguments
+    ///
+    /// * `did` - the DID of the XMTP device installation
+    /// * `name` - the name of the contact bundle variant
+    /// * `value` - the value of the contact bundle
+    /// * `signature` - the signature of the contact bundle
+    #[method(name = "revokeInstallation")]
```
Comment (jac18281828) [inline xps-gateway/src/rpc/api.rs:118]: 🚀

File: xps-gateway/tests/integration_test.rs [hunk 2]
```diff
@@ -1,82 +1,130 @@
...
-use std::{future::Future, time::Duration};
-use tokio::time::timeout as timeout_tokio;
-
-use xps_gateway::{rpc::XpsClient, types::Message, XpsMethods, XpsServer};
-
-const TEST_TIMEOUT: Duration = Duration::from_secs(10);
-pub const SERVER_HOST: &str = "127.0.0.1";
-
-#[cfg(test)]
-mod it {
```
Comment (jac18281828) [inline xps-gateway/tests/integration_test.rs:17]: This change loses the it module?   That is how I am excluding the anvil tests in the pages build.

**Review by jac18281828 — COMMENTED** [review]
This all looks good.   Sorry it took me so long to get back to this as I was fighting some integration test issues today.   It turns out that this change breaks the production deployment.  I think that is in part my fault since I didn't add the production build to the ci.  

I'll create a ticket and assign it to you.

File: xps-gateway/tests/integration_test.rs (hunk 2 above)
Comment (insipx) [reply xps-gateway/tests/integration_test.rs:17]: ah, I can add it back, I was just adding integration tests that were longer, and wanted to remove some of the indentation, didn't realize it was tied to the gh pages build

**Pull request merged**